src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideStreamMemoryBasedAsyncOverrides.cs

```diff
@@ -0,0 +1,208 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using Resx = Microsoft.NetCore.Analyzers.MicrosoftNetCoreAnalyzersResources;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// CA1840: Reports a diagnostic if a class that directly subclasses <see cref="System.IO.Stream"/> overrides 
+    /// <see cref="System.IO.Stream.ReadAsync(byte[], int, int)"/> and/or <see cref="System.IO.Stream.WriteAsync(byte[], int, int)"/>, 
+    /// and does not override the corrasponding memory-based version.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class ProvideStreamMemoryBasedAsyncOverrides : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1840";
+
+        private static readonly LocalizableString s_localizableTitle = CreateResource(nameof(Resx.ProvideStreamMemoryBasedAsyncOverridesTitle));
+        private static readonly LocalizableString s_localizableMessage = CreateResource(nameof(Resx.ProvideStreamMemoryBasedAsyncOverridesMessage));
+        private static readonly LocalizableString s_localizableDescription = CreateResource(nameof(Resx.ProvideStreamMemoryBasedAsyncOverridesDescription));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private const string ReadAsyncName = "ReadAsync";
+        private const string WriteAsyncName = "WriteAsync";
```

Comment:
nit:

```suggestion
        private const string ReadAsyncName = nameof(System.IO.Stream.ReadAsync);
        private const string WriteAsyncName = nameof(System.IO.Stream.WriteAsync);
```

Reply:
Oops, that was sloppy of me. I will fix that here and in other places where I should have used `nameof`.